avalanche/training/plugins/mir.py

```diff
@@ -0,0 +1,190 @@
+            return
+        samples_x, samples_y, samples_tid = next(self.replay_loader)
+        samples_x, samples_y, samples_tid = (
+            samples_x.to(strategy.device),
+            samples_y.to(strategy.device),
+            samples_tid.to(strategy.device),
+        )
+        # Perform the temporary update with current data
+        grad = torch.autograd.grad(
+            strategy.loss, strategy.model.parameters(), retain_graph=True
+        )
+        model_updated = update_temp(
+            strategy.model, grad, strategy.optimizer.param_groups[0]["lr"]
+        )
+        # Selection of the most interfering samples, no grad required
+        # plus we put the model in eval mode so that the additional
+        # forward pass don't influence the batch norm statistics
+        strategy.model.eval()
+        model_updated.eval()
+        with torch.no_grad():
+            _old_red_strategy = strategy._criterion.reduction
+            strategy._criterion.reduction = "none"
+            old_output = avalanche_forward(strategy.model, samples_x, samples_tid)
+            old_loss = strategy._criterion(old_output, samples_y)
+            new_output = avalanche_forward(model_updated, samples_x, samples_tid)
+            new_loss = strategy._criterion(new_output, samples_y)
+            loss_diff = new_loss - old_loss
+            chosen_samples_indexes = torch.argsort(loss_diff)[
+                len(samples_x) - self.batch_size_mem :
+            ]
+            strategy._criterion.reduction = _old_red_strategy
+        strategy.model.train()
+        # Choose the samples and add their loss to the current loss
+        chosen_samples_x, chosen_samples_y, chosen_samples_tid = (
+            samples_x[chosen_samples_indexes],
+            samples_y[chosen_samples_indexes],
+            samples_tid[chosen_samples_indexes],
+        )
+        replay_output = avalanche_forward(
+            strategy.model, chosen_samples_x, chosen_samples_tid
+        )
+        replay_loss = strategy._criterion(replay_output, chosen_samples_y)
+        strategy.loss += replay_loss
+
+    def after_training_exp(self, strategy: "SupervisedTemplate", **kwargs):
+        self.storage_policy.update(strategy, **kwargs)
+        # Exclude classes that were in the last batch
+        buffer = concat_datasets(
+            [
+                self.storage_policy.buffer_groups[key].buffer
+                for key, _ in self.storage_policy.buffer_groups.items()
+                if int(key) not in torch.unique(strategy.mb_y).cpu()
+            ]
+        )
+        if len(buffer) > 0:
+            self.replay_loader = cycle(
+                torch.utils.data.DataLoader(
+                    buffer, batch_size=self.batch_size_mem, shuffle=True
+                )
+            )
+        else:
+            self.replay_loader = None
+
+
+class ReplayLikeMIR(SupervisedPlugin):
+    """
+    Replay plugin but working like MIR for fair comparison
```

Comment:
Can you describe between the default Avalanche replay and this one here in the docstring?

Reply:
@HamedHemati good point. A good tradeoff to avoid refactoring the implementation could be to move the "MIRLikeReplay" to the continual-learning-baselines repository as an OCL baseline, while avalanche will keep the new ReplayPlugin. What do you think?

Reply:
Hey Antonio, yes sure this MIRLikeReplay is just here for the purpose of the comparison so it has nothing to to in avalanche. I will move it to the PR in continual-baseline.

Reply:
I added the test for now. Just to warn you, I tried to stick closer to the setting by suppressing input rotation and cropping, but this only made things worst .. Now I got 34% for MIR and 37% for replay baselines. I don't know what is wrong. I think for the most part it behaves right (after task 1 2 3 it's above normal replay but last task it drops). Maybe it's just some training artifact that I have to fix. Anyway, I guess later discussions about getting the exact same results than in the paper will happen in the continual-learning-baseline PR.

Reply:
Great! as soon as you fix the pep8 errors we can merge the PR. You can check them locally with the command `pycodestyle avalanche/ examples/ tests/ --exclude examples/tvdetection`

Reply:
@AntonioCarta Sorry, I missed the reply to my comment. My point was to implement both `MIRPlugin` and `ReplayLikeMIR`  using the `ReplayPlugin` from Avalanche. As for `MIRLikeReplay`, it should be just a script to replicate results for that strategy by setting the correct arguments, and for `MIRPlugin`, `self.replay_loader` can be replaced with Avalanche's `ReplayPlugin`.

Reply:
@HamedHemati The thing that I had in mind was to implement it like that for now and later change it once the interactive buffer system is on the main branch. I don't know if this answers your question. As for ReplayLikeMIR I think it makes more sense to put it only on the continual-baseline repo since it is not very different from normal replay, I use it more to try and get close to the ER baseline the authors use in their code.